client/Wayland/wlfreerdp.c

```diff
@@ -117,14 +137,37 @@ BOOL wl_post_connect(freerdp* instance)
 	 /* put Wayland data in the context here */
 	context->window = window;
 
-	freerdp_channels_post_connect(instance->context->channels, instance);
+	if (freerdp_channels_post_connect(instance->context->channels, instance))
+		return FALSE;
 
 	wlf_UpdateWindowArea(context, window, 0, 0, gdi->width, gdi->height);
 
 	return TRUE;
 }
 
-BOOL wl_verify_certificate(freerdp* instance, char* subject, char* issuer, char* fingerprint)
+static void wl_post_disconnect(freerdp* instance)
+{
+	wlfContext *context;
+	if (!instance)
+		return;
+
+	if (!instance->context)
+		return;
+
+	context = (wlfContext*) instance->context;
+
+	if (context->display)
+		wlf_DestroyDisplay(context, context->display);
+
+	if (context->input)
+		wlf_DestroyInput(context, context->input);
+
+	gdi_free(instance);
+	if (context->window)
+		wlf_DestroyWindow(context, context->window);
```

Comment:
Although absolutely correct we could leave out the checks for context->display, context->input, context->input in this case since wlf_DestroyDisplay, wlf_DestroyInput respectively wlf_DestroyWindow do proper input validation.
Just an idea but not necessary.